NWChem: Run dipole moment calculation

ahurta92 opened the pull request:
<!-- Thank you for your contribution! -->

## Description
<!-- Provide a brief description of the PR's purpose here. -->
This pull request allows qcng to run properties calculations.  It also includes changes to harvester.py that reads the dipole moments from nwchem output files.  The PR also includes a test_dipole() function.
## Changelog description
<!-- Provide a brief single sentence for the changelog. -->
Allows qcng to run properties calculations and parse dipole moments output.

## Status
<!-- Please `pip install .[lint]; make format` in the base folder. -->
- [x] Code base linted
- [x] Ready to go

loriab commented on qcengine/programs/nwchem/runner.py:
```diff
@@ -238,7 +238,7 @@ def parse_output(
 
         # Normalize the output as a float or list of floats
-        retres = qcvars[f"CURRENT {input_model.driver.upper()}"]
+        retres = qcvars[f"CURRENT ENERGY" ]
```
as-is this will mess up gradient and hessian jobs. want the original line except for properties driver

ahurta92 replied:
Okay I pushed the change.  I added an if statement for the properties driver

dgasmith commented on qcengine/programs/tests/test_nwchem.py:
```diff
@@ -97,3 +97,40 @@ def test_gradient(nh2):
+    assert res["properties"]["calcinfo_nalpha"] == 6
+    assert res["properties"]["calcinfo_nbasis"] == 13
+    # Make sure Dipole Moment and center of charge parsed correctly
+    assert compare_values(.272949872, float(res["extras"]["qcvars"]["TOTAL DIPOLE MOMENT"]), atol=1e-5)
```
Can you add the preceding 0 on these numbers?

dgasmith commented:
Can you rebase this? Test should pass after that.

dgasmith commented:
@ahurta92 Small heads up, we will be cutting a release soon and would love to get this in!

ahurta92 commented:
Okay great,  I rebased and fixed a bug.  It looks like everything was
approved

On Wed, Feb 5, 2020 at 11:51 AM Daniel Smith <notifications@github.com>
wrote:

> @ahurta92 <https://github.com/ahurta92> Small heads up, we will be
> cutting a release soon and would love to get this in!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MolSSI/QCEngine/pull/220?email_source=notifications&email_token=AFP7UTMZYACKRBYXSVQVUQLRBLU73A5CNFSM4KKP4GCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEK4EQ4Q#issuecomment-582502514>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFP7UTMHO3UST32C4TFEUQ3RBLU73ANCNFSM4KKP4GCA>
> .
>

dgasmith commented:
Thanks for patching everything up! @loriab Can you give a final review?

loriab commented on qcengine/programs/nwchem/harvester.py:
```diff
@@ -614,11 +614,103 @@ def harvest_outfile_pass(outtext):
+                + r"[0]"
+                + r"(\d+)",
+                outtext,
+                re.MULTILINE,
```
black and yapf should both be reproached for their handling of regex

loriab commented on qcengine/programs/nwchem/harvester.py:
```diff
@@ -614,11 +614,103 @@ def harvest_outfile_pass(outtext):
+                    psivar["HOMO"] = np.array([round(lumo, 10)])
+                else:
+                    lumo = float(mobj.group(1)) * (10 ** (-1 * float(mobj.group(3))))
+                    psivar["HOMO"] = np.array([round(lumo, 10)])
```
am I missing how the `else` block differs from the `if` block?

ahurta92 replied:
Sorry, my student and I used this for testing.  I just removed the if-else blocks.

ahurta92 replied:
Actually just spoke to my student.  At first, he told me it was just for testing but the reason for the IF else is to detect the sign of the power e.g "2.125011D-01",  I'm changing it again.  Sorry about that

loriab commented on qcengine/programs/nwchem/harvester.py:
```diff
@@ -614,11 +614,103 @@ def harvest_outfile_pass(outtext):
+                    psivar["LUMO"] = np.array([round(lumo, 10)])
+                else:
+                    lumo = float(mobj.group(1)) * (10 ** (-1 * float(mobj.group(3))))
+                    psivar["LUMO"] = np.array([round(lumo, 10)])
```
same if/else query

loriab commented on qcengine/programs/nwchem/harvester.py:
```diff
@@ -614,11 +614,103 @@ def harvest_outfile_pass(outtext):
+        + r"\s+"
+        + r"Y\s+=\s+([+-]?\d+[.]\d+)"
+        + r"\s+"
+        + r"Z\s+=\s+([+-]?\d+[.]\d+)",
```
this is fine if purposeful, but I wanted to make sure you're aware of `NUMBER`, which can simplify reading regex blocks

ahurta92 replied:
I wouldn't say it's purposeful, I can try changing it to simplify the code.  I wasn't aware of `NUMBER`

loriab replied:
thanks, it'll help readability. fine as a future pass.
